pkg/rest/config_upload_test.go

```diff
@@ -424,36 +424,151 @@ func Test_retryReturnContainsHttpErrorIfNotSuccess(t *testing.T) {
+				{400, `epic fail`},
+			},
+			apiKey:      "slo",
+			expectError: true,
+		},
+		{
+			name: "Sends correct param to get all SLOs",
+			expectedQueryParamsPerApiCall: [][]testQueryParams{
+				{
+					{"enabledSlos", "all"},
+				},
+			},
+			expectedApiCalls: 1,
+			serverResponses: []testServerResponse{
+				{200, `{ "values": [ {"id": "1", "name": "name1"} ] }`},
+			},
+			apiKey:      "slo",
+			expectError: false,
+		},
+		{
+			name: "Sends correct parameters for paginated SLO responses",
+			expectedQueryParamsPerApiCall: [][]testQueryParams{
+				{
+					{"enabledSlos", "all"},
+				},
+				{
+					{"nextPageKey", "page42"},
+				},
+				{
+					{"nextPageKey", "page43"},
+				},
+			},
+			expectedApiCalls: 3,
+			serverResponses: []testServerResponse{
+				{200, `{ "nextPageKey": "page42", "values": [ {"id": "1", "name": "name1"} ] }`},
+				{200, `{ "nextPageKey": "page43", "values": [ {"id": "2", "name": "name2"} ] }`},
+				{200, `{ "values": [ {"id": "3", "name": "name3"} ] }`},
+			},
+			apiKey:      "slo",
+			expectError: false,
 		},
 		{
-			expectedQueryParam:      "includeEntityFilterMetricEvents",
-			expectedQueryParamValue: "true",
-			apiKey:                  "anomaly-detection-metrics",
+			name: "Sends correct param to get all anomaly detection metrics",
+			expectedQueryParamsPerApiCall: [][]testQueryParams{
+				{
+					{"includeEntityFilterMetricEvents", "true"},
+				},
+			},
+			expectedApiCalls: 1,
+			serverResponses: []testServerResponse{
+				{200, `{ "values": [ {"id": "1", "name": "name1"} ] }`},
+			},
+			apiKey:      "anomaly-detection-metrics",
+			expectError: false,
 		},
 	}
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
+			apiCalls := 0
 			server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
-				addedQueryParameter := req.URL.Query()[tt.expectedQueryParam]
-				assert.Check(t, addedQueryParameter != nil)
-				assert.Check(t, len(addedQueryParameter) > 0)
-				assert.Equal(t, addedQueryParameter[0], tt.expectedQueryParamValue)
+				if len(tt.expectedQueryParamsPerApiCall) > 0 {
+					params := tt.expectedQueryParamsPerApiCall[apiCalls]
+					for _, param := range params {
+						addedQueryParameter := req.URL.Query()[param.key]
+						assert.Assert(t, addedQueryParameter != nil)
+						assert.Assert(t, len(addedQueryParameter) > 0)
+						assert.Equal(t, addedQueryParameter[0], param.value)
+					}
+				} else {
+					assert.Equal(t, "", req.URL.RawQuery, "expected no query params - but '%s' was sent", req.URL.RawQuery)
+				}
+
+				resp := tt.serverResponses[apiCalls]
+				if resp.statusCode != 200 {
+					http.Error(rw, resp.body, resp.statusCode)
+				} else {
+					_, _ = rw.Write([]byte(resp.body))
+				}
+
+				apiCalls++
+				assert.Check(t, apiCalls <= tt.expectedApiCalls, "expected at most %d API calls to happen, but encountered call %d", tt.expectedApiCalls, apiCalls)
```

Comment:
this might go a bit far into testing the exact behaviour/flow of execution - what do you think?